index.d.ts

```diff
@@ -106,46 +108,34 @@ declare namespace winston {
     rejectionHandlers?: any;
   }
 
-  class Logger extends NodeJSStream.Transform {
-    constructor(options?: LoggerOptions);
+  // Used to dynamically add log levels
+  type LoggerLogFunctionsMap<T extends string> = {
+    [key in T]: LeveledLogMethod;
+  };
+  type LoggerLogEnabledMap<T extends string> = {
+    [key in `is${Capitalize<T>}Enabled`]: () => boolean;
+  };
+
+  interface LoggerBase<T extends string> extends NodeJSStream.Transform {
+    new <T extends string = defaultLevels>(options?: LoggerOptions<T>): Logger<T>;
 
     silent: boolean;
     format: logform.Format;
-    levels: Config.AbstractConfigSetLevels;
-    level: string;
+    levels: Config.AbstractConfigSetLevels<T>;
+    level: T;
     transports: Transport[];
     exceptions: ExceptionHandler;
     rejections: RejectionHandler;
     profilers: object;
     exitOnError: Function | boolean;
     defaultMeta?: any;
 
-    log: LogMethod;
+    log: LogMethod<T>;
     add(transport: Transport): this;
     remove(transport: Transport): this;
     clear(): this;
     close(): this;
 
-    // for cli and npm levels
-    error: LeveledLogMethod;
-    warn: LeveledLogMethod;
-    help: LeveledLogMethod;
-    data: LeveledLogMethod;
-    info: LeveledLogMethod;
-    debug: LeveledLogMethod;
-    prompt: LeveledLogMethod;
-    http: LeveledLogMethod;
-    verbose: LeveledLogMethod;
-    input: LeveledLogMethod;
-    silly: LeveledLogMethod;
```

Comment:
```
  [key in T]: LeveledLogMethod;
```
should declare the respective methods for immediate use.

Might make sense to have the levels with their respective numbers as Type parameter:
```
interface LoggerBase<Levels extends AbstractConfigSetLevels> {
  [key in keyof Levels]: LeveledLogMethod;
}
```

Reply:
mind that custom log levels still break application-insights instrumentation translating them to the correct appinsights severities.

Reply:
@eschablowski  thoughts?